prolif/protein_helper.py

```diff
@@ -0,0 +1,560 @@
+
+        em.AddBond(idx1, idx2, order=order)
+
+    mol = em.GetMol()
+
+    # find lone hydrogens, then attach to the nearest heavy atom
+    _assign_intra_props_lone_H(mol)
+
+    for _, row in reference_block["_chem_comp_atom"].iterrows():
+        nm, arom = row["atom_id"].strip(), row["pdbx_aromatic_flag"].strip()
+        try:
+            idx = nm_2_idx[nm]
+        except KeyError:
+            # todo: could check atom is marked as leaving atom
+            continue
+
+        atom = mol.GetAtomWithIdx(idx)
+        atom.SetIsAromatic(arom == "Y")
+
+    # assign formal charge for a specific atom
+    if fc_special_assign_nm_idx_fc_pair != {}:
+        for (
+            fc_special_assign_nm_idx,
+            fc_to_assign,
+        ) in fc_special_assign_nm_idx_fc_pair.items():
+            atom = mol.GetAtomWithIdx(fc_special_assign_nm_idx)
+            atom.SetFormalCharge(fc_to_assign)
+            logger.debug(
+                f"Assigned {reference_block['name']}'s formal charge {fc_to_assign} "
+                f"on {atom.GetPDBResidueInfo().GetName()}."
+            )
+
+    # sanitize the molecule
+    mol.UpdatePropertyCache()
+    Chem.SanitizeMol(mol)
+
+    return mol, valence
+
+
+def _assign_intra_props_lone_H(mol):
+    additional_bonds = []
+    heavy_atoms = []
+    lone_H = []
+    for atom in mol.GetAtoms():
+        if atom.GetAtomicNum() != 1:
+            heavy_atoms.append(atom.GetIdx())
+            continue
+        if atom.GetBonds():  # if any bonds, we're ok
+            continue
+        lone_H.append(atom.GetIdx())
+    if lone_H:
+        logger.debug(f"found lone hydrogens: {lone_H}")
+        conf = mol.GetConformer()
+        for idx in lone_H:
+            pos = conf.GetAtomPosition(idx)
+            minidx, mindist = -1, float("inf")
+            for idx2 in heavy_atoms:
+                pos2 = conf.GetAtomPosition(idx2)
+                d = pos.Distance(pos2)
+                if d > mindist:
+                    continue
+                minidx, mindist = idx2, d
+
+            if mindist < MAX_AMIDE_LENGTH:
+                logger.debug(
+                    f"attached hydrogen {idx} to heavy atom {minidx} d={mindist}"
+                )
+                additional_bonds.append((idx, minidx))
+    if additional_bonds:
+        em = Chem.EditableMol(mol)
+        for i, j in additional_bonds:
+            em.AddBond(i, j, order=Chem.BondType.SINGLE)
+        mol = em.GetMol()
+
+    return mol
+
+
+# amino acid template
+# ALA, ARG, ASN,
+# ASP (Deprotonated (-1), default), ASH (Neutral (0)),
+# CYS (Neutral (0), default), CYX,
+# GLN, GLY,
+# GLU (Deprotonated (-1), default), GLH (Neutral (0)),
+# HIS (Both NE2 and ND1 protonated, default),
+# HIE (neutral, hydrogen at NE2, might couple to HEME at ND1),
+# HID (neutral, hydrogen at ND1, might couple to HEME at NE2),
+# ILE, LEU, LYS, MET, PRO, PHE, SER, TYR, THR, TRP, VAL
+# HOH (water)
+STANDARD_AA = cif_parser_lite(_STANDARD_AA)
```

Comment:
this should be in the constants file, or at the top of this file if there's a circular import